Integrate NIC hotplug with LiveUpdate rollout strategy

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
VEP tracking issue: https://github.com/kubevirt/enhancements/issues/29.
In KubeVirt v1.5, the namespace admin's permissions for creating VMIM objects were removed by default as a hardening measure.
This broke the NIC hotplug / hotunplug feature that required a manual migration to be performed, following the change to the VM object.

The PR integrates the NIC hotplug feature with the LiveUpdate rollout strategy feature, and enables automatic migration of the VMI following a NIC hotplug/hotunplug.

In order to enable this feature, the following configuration should be made to the KubeVirt CR:
```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  configuration:
    vmRolloutStrategy: LiveUpdate
  workloadUpdateStrategy:
    workloadUpdateMethods:
    - LiveMigrate
```

The implementation enables co-existence with the in-place NIC hotplug / unplug feature that uses the dynamic-networks-controller by giving a grace period for the DNC to perform its actions.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Integrate NIC hotplug with LiveUpdate rollout strategy
```

orelmisan commented:
/uncc @Barakmor1 @fossedihelm @iholder101 @RamLavi

orelmisan commented:
/cc @EdDev @nirdothan

orelmisan commented:
Manually tested with the following KubeVirt configuration:
```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  certificateRotateStrategy: {}
  configuration:
    developerConfiguration:
      featureGates:
      - VMLiveUpdateFeatures
    vmRolloutStrategy: LiveUpdate
  customizeComponents: {}
  imagePullPolicy: IfNotPresent
  workloadUpdateStrategy:
    workloadUpdateMethods:
    - LiveMigrate
```

orelmisan commented:
/pull-kubevirt-e2e-kind-sriov
/pull-kubevirt-e2e-k8s-1.32-sig-network

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/fcca7d172a3d32720dc9d51749cf974984722a40..502b45211e334df06b97fdcfd9378b75de8eca10): Added time-based scenarios to network migration evaluator.

jean-edouard commented on tests/network/hotplug_sriov.go:
```diff
@@ -53,6 +56,23 @@ var _ = Describe(SIG(" SRIOV nic-hotplug", Serial, decorators.SRIOV, func() {
+		updateStrategy := &v1.KubeVirtWorkloadUpdateStrategy{
+			WorkloadUpdateMethods: []v1.WorkloadUpdateMethod{v1.WorkloadUpdateMethodLiveMigrate},
+		}
+		rolloutStrategy := pointer.P(v1.VMRolloutStrategyLiveUpdate)
```
Other tests that set the LiveUpdate rollout strategy add the decorator `VMLiveUpdateRolloutStrategy`. Not sure why honestly, but maybe we want to do the same here.

jean-edouard replied:
> > Other tests that set the LiveUpdate rollout strategy add the decorator `VMLiveUpdateRolloutStrategy`. Not sure why honestly, but maybe we want to do the same here.
> 
> Is it used anywhere in our CI?

It doesn't appear to be... Found this though: https://github.com/kubevirt/kubevirt/pull/13091/files#r1810163792

jean-edouard replied:
FYI @fossedihelm

orelmisan replied:
If it doesn't do anything, maybe we should consider removing it in a follow-up.

jean-edouard reviewed (approved):
Looks great! Not /approving yet to give time for sig-network reviews.

oshoval commented on pkg/virt-controller/watch/vmi/lifecycle.go:
```diff
@@ -966,6 +969,56 @@ func (c *Controller) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
+		return
+	}
+
+	if existingCondition != nil && existingCondition.Status == k8sv1.ConditionTrue {
```
once `existingCondition.Status == k8sv1.ConditionTrue`
it can't go back to `result == k8sv1.ConditionUnknown` right ?

if so, would it make sense to move it before the switch ?
(i do understand the current method is safer wrt to its own functionality, but if we don't expect such case, maybe it is better to be explicit and not keep potential unpredictable states)

oshoval reviewed (commented):
Thanks, nice
partial review

oshoval commented on pkg/virt-controller/watch/vmi/lifecycle.go:
```diff
@@ -966,6 +969,56 @@ func (c *Controller) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
 }
 
+func (c *Controller) syncAutoMigration(vmi *virtv1.VirtualMachineInstance) {
+	const pendingMigrationReEvalPeriod = 10 * time.Second
```
nit: also worth to move near usage please

orelmisan replied:
> Don't you want to have this plugable, to be able to manipulate in testing?

It is not tested in the scope of the unit tests, because of the nature of the queue's `AddAfter` method.
However, it is tested as part of the e2e tests.

oshoval commented on tests/network/hotplug_bridge.go:
```diff
@@ -231,7 +255,24 @@ var _ = Describe(SIG("bridge nic-hotplug", func() {
-var _ = Describe(SIG("bridge nic-hotunplug", func() {
+var _ = Describe(SIG("bridge nic-hotunplug", Serial, func() {
+	BeforeEach(func() {
+		virtClient := kubevirt.Client()
+		originalKv := libkubevirt.GetCurrentKv(virtClient)
+		updateStrategy := &v1.KubeVirtWorkloadUpdateStrategy{
+			WorkloadUpdateMethods: []v1.WorkloadUpdateMethod{v1.WorkloadUpdateMethodLiveMigrate},
+		}
+		rolloutStrategy := pointer.P(v1.VMRolloutStrategyLiveUpdate)
+		patchWorkloadUpdateMethodAndRolloutStrategy(originalKv.Name, virtClient, updateStrategy, rolloutStrategy)
+
+		currentKv := libkubevirt.GetCurrentKv(virtClient)
+		config.WaitForConfigToBePropagatedToComponent(
+			"kubevirt.io=virt-controller",
+			currentKv.ResourceVersion,
+			config.ExpectResourceVersionToBeLessEqualThanConfigVersion,
+			time.Minute)
+	})
```
this blocks repeats right ?
please consider adding a function
it also repeats on next commit of this PR `tests/network/hotplug_sriov.go`

it also might repeat on other files (`tests/hotplug/instancetype.go` unless not exactly the same)
anyhow if it is messy that both will use the same, we can just start at least with the repetitions in this PR

orelmisan replied:
> this blocks repeats right ? please consider adding a function it also repeats on next commit of this PR `tests/network/hotplug_sriov.go`
> 
> it also might repeat on other files (`tests/hotplug/instancetype.go` unless not exactly the same) anyhow if it is messy that both will use the same, we can just start at least with the repetitions in this PR

This code was adjusted from the CPU hotplug tests.
It makes sure the LiveUpdate Rollout strategy is set prior to running the tests.

I think we could consider it for a follow-up.

oshoval replied:
well you are the maintainer, but i think this is an example it scream repetitions
if the helper function is problematic as it cross package then fine, but what about simple one in the same file / package ?

oshoval commented on tests/network/hotplug_sriov.go:
```diff
@@ -87,9 +107,28 @@ var _ = Describe(SIG(" SRIOV nic-hotplug", Serial, decorators.SRIOV, func() {
-			migration := libmigration.New(hotPluggedVMI.Name, hotPluggedVMI.Namespace)
-			migrationUID := libmigration.RunMigrationAndExpectToCompleteWithDefaultTimeout(kubevirt.Client(), migration)
-			libmigration.ConfirmVMIPostMigration(kubevirt.Client(), hotPluggedVMI, migrationUID)
+			virtClient := kubevirt.Client()
+
+			By("Waiting for MigrationRequired condition to appear")
+			Eventually(matcher.ThisVMI(hotPluggedVMI), 1*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceMigrationRequired))
+
+			By("Ensuring live-migration started")
+			var migration *v1.VirtualMachineInstanceMigration
+			Eventually(func() bool {
+				migrations, err := virtClient.VirtualMachineInstanceMigration(hotPluggedVMI.Namespace).List(context.Background(), metav1.ListOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				for _, mig := range migrations.Items {
+					if mig.Spec.VMIName == hotPluggedVMI.Name {
+						migration = mig.DeepCopy()
+						return true
+					}
+				}
+				return false
+			}, 30*time.Second, time.Second).Should(BeTrue())
+
+			libmigration.ExpectMigrationToSucceedWithDefaultTimeout(virtClient, migration)
+			libmigration.ConfirmVMIPostMigration(kubevirt.Client(), hotPluggedVMI, migration)
```
this code repeats in previous commit, can we use an helper function for those two ?

orelmisan replied:
I personally am not a fan of helper functions containing gomega statements, as they don't point you at the right direction when errors occur.

oshoval replied:
personally i think it is a bad idea, the stack depth is known at given time and doesnt change a lot

orelmisan commented on pkg/network/migration/evaluator.go:
```diff
@@ -0,0 +1,124 @@
+
+	switch result {
+	case notRequired:
+		return k8scorev1.ConditionUnknown
```
> Isn't this an abuse of `ConditionUnknown`? We know the condition. Is this hack already used elsewhere in the project?

Could you please elaborate why do you think it is an abuse / hack?

There is a need for the SIG-specific logic to express its desire to:
1. Not require a migration.
2. Require an immediate migration.
3. Require a pending migration.

Since the SIG-specific and the general logics are decoupled, there is a need to have a type shared by both.
I think the `ConditionUnknown` is a reasonable.
In case you have a better alternative - please raise it.

lyarwood replied:
I agree that returning `ConditionUnknown` seems wrong when we know that a migration is not required but return `ConditionFalse` when we know one might eventually be required. What's stopping us from introducing our own conditions for this instead of these core k8s conditions?

lyarwood replied:
It's both really, for the interface it seems weird to have an implementations return `ConditionUnknown` when a migration is not required. Would `nil` be cleaner?

To also express that a migration is required but pending action by a third party with `VirtualMachineInstanceMigrationRequired` `ConditionFalse` on the VMI itself also doesn't seem right. Personally I'd break that out into another explicit `VirtualMachineInstanceConditionType` that you can assign `ConditionTrue`. I appreciate that's more work but the resulting conditions should be easier to understand IMHO.

Either way would it be possible to add some docs regarding implementation expectations alongside the interface?

orelmisan replied:
> It's both really, for the interface it seems weird to have an implementations return `ConditionUnknown` when a migration is not required. Would `nil` be cleaner?

@EdDev, @vladikr WDYT?

> To also express that a migration is required but pending action by a third party with `VirtualMachineInstanceMigrationRequired` `ConditionFalse` on the VMI itself also doesn't seem right. Personally I'd break that out into another explicit `VirtualMachineInstanceConditionType` that you can assign `ConditionTrue`. I appreciate that's more work but the resulting conditions should be easier to understand IMHO.

Aren't the reasons that accompany the condition clear enough in your opinion?

> Either way would it be possible to add some docs regarding implementation expectations alongside the interface?

Yes, it will be added to the commit message after we will come to an agreement about the interface.

lyarwood replied:
> Aren't the reasons that accompany the condition clear enough in your opinion?

That is true they do help but it still feels awkward to express the pending state with a false condition. I'm not going to die on this hill but details like this are why I suggested breaking this work out into a separate enhancement and implementation.

> Yes, it will be added to the commit message after we will come to an agreement about the interface.

A code comment alongside the interface would be more useful for folks providing additional implementations in the future.

vladikr commented on pkg/network/migration/evaluator.go:
```diff
@@ -0,0 +1,124 @@
+
+func (e Evaluator) Evaluate(vmi *v1.VirtualMachineInstance) k8scorev1.ConditionStatus {
+	result := shouldVMIBeMarkedForAutoMigration(
+		vmi.Spec.Domain.Devices.Interfaces,
+		vmi.Spec.Networks,
+		vmi.Status.Interfaces,
```
Are we sure that all of these fields can't be `nil` at this point?

orelmisan replied:
The code is expected to work even if all three slices are `nil`.
There is a unit test for it in `pkg/network/migration/evaluator_test.go`:
> Evaluator Should not require migration when no networks are specified.

vladikr replied:
Thanks, you're right

vladikr commented on pkg/virt-controller/watch/vmi/lifecycle.go:
```diff
@@ -428,6 +429,8 @@ func (c *Controller) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8sv1
 			c.syncVolumesUpdate(vmiCopy)
 		}
 
+		c.syncAutoMigration(vmiCopy)
```
@orelmisan can we do something about this name? I think it should contain the word "Condition" but also isn't it networking related?

orelmisan replied:
How about `syncMigrationRequiredCondition`?

Currently, SIG network is its only customer, but it could be used to serve other SIGs by replacing the SIG-network migration evaluator with an evaluator that can combine the results of multiple SIGs/flows.

vladikr replied:
Thanks, yes, that would be fine.

orelmisan commented on pkg/virt-controller/watch/migration/migration.go:
```diff
@@ -681,7 +681,8 @@ func (c *Controller) processMigrationPhase(
 			!vmiConditionManager.HasCondition(vmi, virtv1.VirtualMachineInstanceVCPUChange) &&
-			!vmiConditionManager.HasConditionWithStatus(vmi, virtv1.VirtualMachineInstanceMemoryChange, k8sv1.ConditionTrue) {
+			!vmiConditionManager.HasConditionWithStatus(vmi, virtv1.VirtualMachineInstanceMemoryChange, k8sv1.ConditionTrue) &&
+			!vmiConditionManager.HasConditionWithStatus(vmi, virtv1.VirtualMachineInstanceMigrationRequired, k8sv1.ConditionTrue) {
```
Please see https://github.com/kubevirt/enhancements/blob/a06daa137f659a6404683a1bd0975bb13528690a/veps/sig-network/ns-admin-nichotplug.md?plain=1#L69.

lyarwood reviewed (commented):
Partial review but Looks really nice thus far, I do have a concern with the use of `Condition{Unknown,True,False}` instead of using something more explicit for the false and pending condition returns from the evaluator.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/ad716aad30578cbef94d577ff395e194915765e9..3bf3ace80ef12e3c964393cb897bce71afd8bc21): Addressed @vladikr's [comment](https://github.com/kubevirt/kubevirt/pull/14259#discussion_r2139756568).

vladikr commented on pkg/virt-controller/watch/vmi/lifecycle.go:
```diff
@@ -966,6 +969,56 @@ func (c *Controller) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
 	}
 }
 
+func (c *Controller) syncMigrationRequiredCondition(vmi *virtv1.VirtualMachineInstance) {
```
yes, I think my point is that if this function decides to enqueue, we will not get to the [updateStatus](https://github.com/kubevirt/kubevirt/pull/14259/files/3bf3ace80ef12e3c964393cb897bce71afd8bc21#diff-513cdb230d096b30280705ead0601d957bbdd658cf85a1017a4ea7759763feadR447) on line 447
Am I right?

orelmisan replied:
This method does not change the current flow.
If something happened previously - it will still happen.

vladikr commented on pkg/virt-controller/watch/vmi/lifecycle.go:
```diff
@@ -966,6 +969,56 @@ func (c *Controller) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
+	}
+
+	key, _ := controller.KeyFunc(vmi)
+	c.Queue.AddAfter(key, pendingMigrationReEvalPeriod)
```
@orelmisan shouldn't we decide to give up (not to reenque) at some point? 
Is there a maximum amount of with we should wait?

orelmisan replied:
The Evaluator in the network package will set the condition to true after about 15 seconds (could be less if dynamic-networks-controller is involved, or a bit more if the VMI is handled after the initial 15 seconds).

In any other case - the method does not re-enqueue.

orelmisan replied:
A comment was added above the re-enqueue logic to explain the situation.

> I would prefer it at the function level, to reduce the chance for rotten comments.

I missed this sentence, apologies.

orelmisan commented on pkg/virt-controller/watch/vmi/lifecycle.go:
```diff
@@ -966,6 +969,56 @@ func (c *Controller) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
+	}
+
+	key, _ := controller.KeyFunc(vmi)
+	c.Queue.AddAfter(key, pendingMigrationReEvalPeriod)
```
Yes.

oshoval commented:
/lgtm

EdDev commented on tests/network/hotplug_bridge.go:
```diff
@@ -339,10 +380,28 @@ func removeInterface(vm *v1.VirtualMachine, name string) error {
+		Eventually(matcher.ThisVMI(vmi), 1*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceMigrationRequired))
+
+		By("Ensuring live-migration started")
+		var migration *v1.VirtualMachineInstanceMigration
```
I would prefer to check the migration started on the VMI itself and not the VMIM. Is that possible and makes sense?

orelmisan replied:
The test still requires the VMIM (even before this change) in order to track the migration completion.

orelmisan replied:
Hi, In general it makes sense, I will have to test it in a follow up.

The change proposed in this commit was meant to fetch the VMI in order to make sure it actually ended.

EdDev commented on tests/network/hotplug_bridge.go:
```diff
@@ -339,10 +380,28 @@ func removeInterface(vm *v1.VirtualMachine, name string) error {
+
+			for _, mig := range migrations.Items {
+				if mig.Spec.VMIName == vmi.Name {
+					migration = mig.DeepCopy()
```
IMO the migration should be returned and not a boolean. Failure of a boolean says nothing in addition to its nasty look.

orelmisan replied:
A pointer to VMIM is now returned and asserted to to be nil.

EdDev reviewed (commented):
Thank you, looks very clean and nice.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/15462fc2858d318202fdd3b1fa32b49511ddfcee..1482a22cc18b850e585c0962c61718eaf3a07083): Addressed review comments.

orelmisan commented on pkg/network/migration/evaluator.go:
```diff
@@ -0,0 +1,124 @@
+	return notRequired
+}
+
+func getMigrationRequiredCondition(conditions []v1.VirtualMachineInstanceCondition) *v1.VirtualMachineInstanceCondition {
```
Done.

orelmisan commented on pkg/network/migration/evaluator_test.go:
```diff
@@ -0,0 +1,208 @@
+		),
+	)
+
+	It("Should require an immediate migration when a secondary iface using SR-IOV binding is hotplugged", func() {
```
SR-IOV hot-unplug is currently not supported and is blocked on the validating webhook.
The reason was probably the requirement to manually migrate the VM following the detachment of the SR-IOV NIC in order to free the resources from the virt-launcher pod.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/1482a22cc18b850e585c0962c61718eaf3a07083..1c0649486216225f601c2428d47e9928b2f1f54d): Addressed review comments.

orelmisan commented on tests/network/hotplug_bridge.go:
```diff
@@ -339,10 +380,28 @@ func removeInterface(vm *v1.VirtualMachine, name string) error {
+				}
+			}
+			return false
+		}, 30*time.Second, time.Second).Should(BeTrue())
```
Done.

vladikr commented:
/test all

orelmisan commented:
/retest-required

orelmisan commented:
The flakes were in unrelated lanes.
/hold cancel
